### PR TITLE
[MRG] Add weak OT solver

### PR DESCRIPTION
**Pull request by rflamary**
## Types of changes

* Add weak OT solver
* Add tests for solver and backends
* Add example for weak OT in doc

## PR checklist

- [x] I have read the [**CONTRIBUTING**](CONTRIBUTING.md) document.
- [x] The documentation is up-to-date with the changes I made (check build artifacts).
- [x] All tests passed, and additional code has been **covered with new tests**.
- [x] I have added the PR and Issue fix to the [**RELEASES.md**](RELEASES.md) file.

### CODE REVIEW

File: ot/weak.py [hunk 1]
```diff
@@ -0,0 +1,123 @@
...
+    ot.optim.cg : General regularized OT
+    """
+
+    nx = get_backend(Xa, Xb)
+
+    Xa2 = nx.to_numpy(Xa)
+    Xb2 = nx.to_numpy(Xb)
+
+    if a is None:
+        a2 = nx.ones((Xa.shape[0]), type_as=Xa) / Xa.shape[0]
```
Comment (ncourty) [inline ot/weak.py:94]: type_as should be related to Xa2 or directly in numpy ?

File: test/test_weak.py [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+
+
+def test_weak_ot_bakends(nx):
+    # test weak ot solver for different backends
+    n = 50
+    rng = np.random.RandomState(0)
+
+    xs = rng.randn(n, 2)
+    xt = rng.randn(n, 2)
+    u = ot.utils.unif(n)
```
Comment (ncourty) [inline test/test_weak.py:44]: makes me think there is no type_as for unif ?

**Review by ncourty — CHANGES_REQUESTED** [review]
thx for the nice work rémi !

File: ot/weak.py [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+# License: MIT License
+
+from .backend import get_backend
+from .optim import cg
+import numpy as np
+
+__all__ = ['weak_optimal_transport']
+
+
+def weak_optimal_transport(Xa, Xb, a=None, b=None, verbose=False, log=False, G0=None, **kwargs):
```
[the remark this answers is not in the document]
Comment (rflamary) [reply ot/weak.py:16]: the API of this function is closer to empirical sinkhorn
https://pythonot.github.io/gen_modules/ot.bregman.html?highlight=empirical_sinkhorn#ot.bregman.empirical_sinkhorn

where the weights are usually uniform so this order makes more sens

File: ot/weak.py (hunk 1 above)
Comment (rflamary) [reply ot/weak.py:94]: good catch

File: test/test_weak.py (hunk 2 above)
Comment (rflamary) [reply test/test_weak.py:44]: added even though it is the only function in teh public API with this parameter (type cannot be inferred from integer)

File: ot/weak.py [hunk 4]
```diff
@@ -0,0 +1,123 @@
...
+
+             \gamma \geq 0
+
+    where :
+
+    - :math:`X_a` :math:`X_b`  are the sample matrices.
+    - :math:`\mathbf{a}` and :math:`\mathbf{b}` are the sample weights
+
+
+    .. note:: This function is backend-compatible and will work on arrays
```
Comment (ncourty) [inline ot/weak.py:35]: yes  but maybe you can add a note about  the fact that all  the computations happen on the CPU in numpy mode

**Review by ncourty — APPROVED** [review]

**Pull request merged**